Don't create new connections during test transaction setup

wjordan opened the pull request:
### Summary

This PR prevents new connections from being eagerly created on all connection pools during test-transaction setup.
Fixes #27581.

### Other Information

The current implementation's attempt to consistently apply transactions to all present and future connections within a transactional test suffers from an issue where the transactional test setup eagerly checks out connections on all connection pools, including those that do not currently contain any connections.

This PR fixes this issue via the following changes:

- Renames the existing `!connection.active_record` notification from #20818, which is triggered when a connection pool (not an individual connection) is newly established, to `!connection_pool.active_record`.
- Creates a new `!connection_pool.active_record` notification hook that fires when a connection is checked out (in the `checkout_and_verify` method), alongside the existing `:checkout` callback runner.
- Updates the existing notification subscription to accept a connection rather than a connection-pool specification when invoked.
- Updates `enlist_fixture_connections` to only return the currently-active connection from each pool, rather than actively checking out a new connection on a pool without any active connection.

rafaelfranca commented:
LGTM. @matthewd WDYT?

eileencodes commented:
Looks like this is causing some build failures in Active Record. Can you look into it @wjordan?

wjordan commented:
I've been looking into into the failures, but might take me a bit longer to sort out. Something in the new connection notification is causing some existing connection-pool tests to timeout/deadlock, though I'm not sure why at this point.

wjordan commented:
OK, I tracked down what was causing the test failures. Several tests within `ConnectionPoolTest` depend on expectations that fetching a connection on a separate thread attempts to checkout a separate connection within the connection pool. (See for example, [`test_checkout_behavior`](https://github.com/rails/rails/blob/5146d8e2081c0e2e7b0a90f323994ba650fe5562/activerecord/test/cases/connection_pool_test.rb#L185-L204).) Since the `lock_thread` implementation added in #28083 is now applied to test transactions, when `lock_thread = true` is set on a transactional test and applied to a connection pool, fetching a connection on a separate thread returns the same connection, so the above expectation is no longer met.

The reason these tests didn't fail until this PR is because in the connection pool tests, [a new connection pool is created](https://github.com/rails/rails/blob/5146d8e2081c0e2e7b0a90f323994ba650fe5562/activerecord/test/cases/connection_pool_test.rb#L12-L13) without any connections until after the test has already begun, so the connections within the test were not previously wrapped by the thread-locked transaction. This PR intentionally wraps all newly-created connections with a thread-locked transaction, causing the test failures to surface.

Considering some Rails tests expect the old connection-fetching behavior, it's safe to assume that there are additional tests out in the wild that probably expect the old behavior as well and will break in similar ways with this change. (I know of [one pattern](http://stackoverflow.com/a/11675647/2518355) already, where a new thread is created to run DDL queries on a separate DB connection to avoid an implicit commit of the test transaction, that would break with this change.) This raises the question how careful we would like to to be rolling out this slightly altered behavior.

I can think of three ways to proceed:

1. Update the failing tests directly (e.g., replacing `pool.connection` implicit checkout with `pool.checkout`), and possibly adding a release note about the behavior change. (I'll make a commit updating this PR along these lines, so at least the tests will pass.)
1. Make the thread-locking test behavior opt-in or otherwise backwards-compatible with existing behavior through a flag on the test case that could be enabled by default for system tests, and manually added to any other test that requires threads to be locked to a single connection (e.g., following current conventions, set `self.lock_threads = true` in the TestCase class and/or `self.lock_threads :method` to thread-lock specific tests)
1. Close this PR until further notice

Thoughts @eileencodes or others?

wjordan commented:
Updated the PR with fixes to `ConnectionPoolTest`. In order to account for the issues in the comment above, I updated the signatures of `#connection` and `#with_connection` to accept an optional `owner_thread` argument (this argument was already present in `#release_connection`), so the default behavior of lock_thread could be optionally overridden (e.g., `Thread.new { pool.with_connection(Thread.current) {} }` approximates the behavior before lock-thread was added).

wjordan committed:
Don't create new connections during test transaction setup
Begin test transactions for new connections in existing pools.
Allow `#connection` and `#with_connection` to accept an `owner_thread`
argument, which overrides the default lock_thread behavior.
Fixes #27581.

matthewd commented on activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:
```diff
@@ -803,7 +803,12 @@ def checkout_new_connection
         def checkout_and_verify(c)
           c._run_checkout_callbacks do
-            c.verify!
+            ActiveSupport::Notifications.instrumenter.instrument(
```
Can we use the existing checkout callbacks ☝🏻 instead of adding this instrumentation?

wjordan replied:
Certainly. I had considered that but decided to use the instrumentation here to match what was already being used for this feature. Is there any significant difference between callbacks vs. notifications, or is it more of an arbitrary choice, and here just a matter of not needing to use both at the same time?

matthewd commented on activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:
```diff
@@ -370,8 +370,8 @@ def lock_thread=(lock_thread)
       # held in a cache keyed by a thread.
-      def connection
-        @thread_cached_conns[connection_cache_key(@lock_thread || Thread.current)] ||= checkout
+      def connection(owner_thread = @lock_thread || Thread.current)
```
I don't like this.. it promotes what's supposed to be a hidden "jam everything, just trust me" switch to being a central component of the class's main public API.

wjordan replied:
Yeah, I'd like some other ideas on how to proceed without affecting the public API if possible. As [noted above](https://github.com/rails/rails/pull/28207#issuecomment-283217928), the underlying issue is that some of the existing [connection pool tests](https://github.com/rails/rails/pull/28207/files#diff-4dec8806a210fe9ff8021207675d4f7c) (likely some end-user tests as well) depend upon the original behavior where threads are not locked to a single connection, so in order to get the tests to pass with minimal changes I chose to modify the public API to allow the new behavior to be optionally overridden. Part of my rationale for choosing this approach was that `#release_connection` already contained `owner_thread` in its public API, so this would just update `#connection` and `#with_connection` to match.

If we don't modify `#connection` and `#with_connection` to match the existing `#release_connection` signature, then we'll need some other way of allowing the lock-thread behavior to be overridden for tests that break on it. Perhaps a switch at the test-case-class or test-method level to mark specific tests to either run outside of a test-transaction entirely, and/or disable the transaction's thread-locking behavior for that test/test-case?

matthewd commented on activerecord/lib/active_record/fixtures.rb:
```diff
@@ -975,20 +975,12 @@ def setup_fixtures(config = ActiveRecord::Base)
-            end
+          if connection && !@fixture_connections.include?(connection)
+            connection.begin_transaction joinable: false
+            connection.pool.lock_thread = true
```
This seems dangerous: it could lead to us locking on any random thread's connection. (Actually already true, now that I read it.. but the change to *when* this occurs seems to make it much more likely to misbehave.)

matthewd commented on activerecord/lib/active_record/fixtures.rb:
```diff
@@ -1020,7 +1012,7 @@ def teardown_fixtures
 
     def enlist_fixture_connections
-      ActiveRecord::Base.connection_handler.connection_pool_list.map(&:connection)
+      ActiveRecord::Base.connection_handler.connection_pool_list.map(&:active_connection?).compact
```
We shouldn't depend on the value of `active_connection?` beyond its truthiness.
